instrumentation/servlet/servlet-common/library/src/main/java/io/opentelemetry/instrumentation/servlet/BaseServletHelper.java

```diff
@@ -0,0 +1,128 @@
+      instrumenter;
+  protected final ServletAccessor<REQUEST, RESPONSE> accessor;
+  private final ServletSpanNameProvider<REQUEST> spanNameProvider;
+
+  protected BaseServletHelper(
+      Instrumenter<ServletRequestContext<REQUEST>, ServletResponseContext<RESPONSE>> instrumenter,
+      ServletAccessor<REQUEST, RESPONSE> accessor) {
+    this.instrumenter = instrumenter;
+    this.accessor = accessor;
+    this.spanNameProvider = new ServletSpanNameProvider<>(accessor);
+  }
+
+  protected Context startSpan(
+      Context parentContext,
+      ServletRequestContext<REQUEST> requestContext,
+      ServerSpanNaming.Source namingSource) {
+    Context context = instrumenter.start(parentContext, requestContext);
+
+    REQUEST request = requestContext.request();
+    SpanContext spanContext = Span.fromContext(context).getSpanContext();
+    // we do this e.g. so that servlet containers can use these values in their access logs
+    // TODO: These are only available when using servlet instrumentation or when server
+    //  instrumentation extends servlet instrumentation e.g. jetty. Either remove or make sure they
+    //  also work on tomcat and wildfly.
+    accessor.setRequestAttribute(request, "trace_id", spanContext.getTraceId());
+    accessor.setRequestAttribute(request, "span_id", spanContext.getSpanId());
+
+    context = ServerSpanNaming.init(context, namingSource);
+    context = addServletContextPath(context, request);
+    context = customizeContext(context, request);
+
+    attachServerContext(context, request);
+
+    return context;
+  }
+
+  /** Override in subclass to customize context that is returned by {@code startSpan}. */
+  protected Context customizeContext(Context context, REQUEST request) {
+    return context;
+  }
+
+  private Context addServletContextPath(Context context, REQUEST request) {
+    String contextPath = accessor.getRequestContextPath(request);
+    if (contextPath != null && !contextPath.isEmpty() && !contextPath.equals("/")) {
+      return context.with(ServletContextPath.CONTEXT_KEY, contextPath);
+    }
+    return context;
+  }
+
+  public Context getServerContext(REQUEST request) {
+    Object context = accessor.getRequestAttribute(request, HttpServerTracer.CONTEXT_ATTRIBUTE);
+    return context instanceof Context ? (Context) context : null;
+  }
+
+  protected void attachServerContext(Context context, REQUEST request) {
+    accessor.setRequestAttribute(request, HttpServerTracer.CONTEXT_ATTRIBUTE, context);
+  }
+
+  public void recordException(Context context, Throwable throwable) {
+    AppServerBridge.recordException(context, throwable);
+  }
+
+  /**
+   * When server spans are managed by app server instrumentation we need to add context path of
+   * current request to context if it isn't already added. Servlet instrumentation adds it when it
+   * starts server span.
+   */
+  public Context updateContext(Context context, REQUEST request) {
```

Comment:
Could this method be merged with `addServletContextPath()`? Checking if current context already contains `ServletContextPath.CONTEXT_KEY` shouldn't probably hurt in `startSpan()`

Reply:
I'll try to do this in a separate pr along with making `ServletContextPath. CONTEXT_KEY` private.